networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BlocksByRangeListenerWrapper.java

```diff
@@ -0,0 +1,111 @@
+
+import java.util.Optional;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.networking.eth2.rpc.core.ResponseStreamListener;
+import tech.pegasys.teku.networking.p2p.peer.Peer;
+
+public class BlocksByRangeListenerWrapper implements ResponseStreamListener<SignedBeaconBlock> {
+
+  private final Peer peer;
+  private final ResponseStreamListener<SignedBeaconBlock> blockResponseListener;
+  private final UInt64 startSlot;
+  private final UInt64 endSlot;
+  private final UInt64 step;
+
+  private Optional<Bytes32> maybeRootOfLastBlock = Optional.empty();
+  private Optional<UInt64> maybeSlotOfLastBlock = Optional.empty();
+
+  public BlocksByRangeListenerWrapper(
+      Peer peer,
+      ResponseStreamListener<SignedBeaconBlock> blockResponseListener,
+      UInt64 startSlot,
+      UInt64 count,
+      UInt64 step) {
+    this.peer = peer;
+    this.blockResponseListener = blockResponseListener;
+    this.startSlot = startSlot;
+    this.step = step;
+    this.endSlot = startSlot.plus(step.times(count));
+  }
+
+  @Override
+  public SafeFuture<?> onResponse(SignedBeaconBlock response) {
+    return SafeFuture.of(
+        () -> {
+          UInt64 blockSlot = response.getSlot();
+          if (!blockSlotIsInRange(blockSlot)) {
+            throw new BlocksByRangeResponseInvalidResponseException(peer, BLOCK_SLOT_NOT_IN_RANGE);
+          }
+
+          if (!blockSlotMatchesTheStep(blockSlot)) {
+            throw new BlocksByRangeResponseInvalidResponseException(
+                peer, BLOCK_SLOT_DOES_NOT_MATCH_STEP);
+          }
+
+          if (!blockSlotGreaterThanPreviousBlockSlot(blockSlot)) {
+            throw new BlocksByRangeResponseInvalidResponseException(
+                peer, BLOCK_SLOT_NOT_GREATER_THAN_PREVIOUS_BLOCK_SLOT);
+          }
+
+          if (!blockParentRootMatches(response.getParent_root())) {
+            throw new BlocksByRangeResponseInvalidResponseException(
+                peer, BLOCK_PARENT_ROOT_DOES_NOT_MATCH);
+          }
+
+          maybeSlotOfLastBlock = Optional.of(blockSlot);
+          maybeRootOfLastBlock = Optional.of(response.getRoot());
+          return blockResponseListener.onResponse(response);
+        });
+  }
+
+  private boolean blockSlotIsInRange(UInt64 blockSlot) {
+    return blockSlot.compareTo(startSlot) >= 0 && blockSlot.compareTo(endSlot) <= 0;
+  }
+
+  private boolean blockSlotMatchesTheStep(UInt64 blockSlot) {
+    return blockSlot.minus(startSlot).mod(step).equals(UInt64.ZERO);
+  }
+
+  private boolean blockSlotGreaterThanPreviousBlockSlot(UInt64 blockSlot) {
+    if (maybeSlotOfLastBlock.isEmpty()) {
+      return true;
+    }
+
+    UInt64 lastBlockSlot = maybeSlotOfLastBlock.get();
+    return blockSlot.compareTo(lastBlockSlot) > 0;
```

Comment:
(nit / optional) We've got some nice comparison utilities now!:
```suggestion
    return blockSlot.isGreaterThan(lastBlockSlot);
```

Reply:
done.